pkg/k8sapi/discovery.go

```diff
@@ -246,21 +246,13 @@ func (d *Controller) handlePodUpdate(key string) error {
 	if d.myNodeName == pod.Spec.NodeName && !pod.Spec.HostNetwork {
 		d.workerPodsLock.Lock()
 		defer d.workerPodsLock.Unlock()
-		var containerID string
-		if len(pod.Status.ContainerStatuses) > 0 && pod.Status.ContainerStatuses[0].ContainerID != "" {
-			containerID = pod.Status.ContainerStatuses[0].ContainerID
-			log.Debugf("Found pod %s with container ID: %s", podName, containerID)
-		} else {
-			log.Debugf("No container ID found for %s", podName)
-		}
 
 		log.Tracef("Update for pod %s: %+v, %+v", podName, pod.Status, pod.Spec)
 
 		// Save pod info
 		d.workerPods[key] = &K8SPodInfo{
 			Name:      podName,
 			Namespace: pod.GetNamespace(),
-			Container: containerID,
```

Comment:
Can the Pod's `K8S_POD_INFRA_CONTAINER_ID` change ever? If so, we may need to update the `K8SPodInfo.Sandbox` attribute accordingly.

Reply:
I don't think we care about the `Sandbox` field here, specifically. The only use of this data in IPAMD is during `nodeInit`, where we extract a copy and then promptly overwrite the `Sandbox` field with data from the CRI. I'm pretty sure we never look at it again, I don't know why we even keep the watch active. (Per https://github.com/aws/amazon-vpc-cni-k8s/pull/714#issuecomment-556457910, there's really no reason for the K8S Pod API to be involved even as a one-shot.)

More generally, the sandbox ID for a Pod **can** change (more on that later), but the CNI/CRI side of things doesn't care. The `ipamd/datastore` `PodKey` store includes the sandbox ID, so we'll treat a sandbox change as a completely unrelated pod. It'll look exactly like one pod getting shut down and another pod getting started up. The fact that they were "the same pod" in some sense doesn't affect anything.

As to how a sandbox ID for a pod can change:

The `key` value here comes from `MetaNamespaceKeyFunc`, which just returns a `namespace/name` string. The sandbox ID associated with that can very easily change -- restarting a pod that's part of a StatefulSet will produce a new pod with the same namespace and name, but a different sandbox (and a different UID).

A less obvious way it can happen is if the sandbox gets re-created. This can happen if the sandbox process gets killed, as well as in some more obscure scenarios which I'm not sure how to trigger (e.g. pod spec changes, which are normally immutable but there's code to handle anyway). In that case, you'll get the same name, namespace, and UID; the only change is the `Attempt` field in the CRI `PodSandboxMetadata` structure, which *is not exposed* in the Kubernetes API.

Because information about the sandbox is generally not exposed through the API, the only observable effects of a sandbox re-creation are that all of the normal containers restart. There's no way to distinguish that the cause of the restart was a sandbox change.

If we treat the CNI/CRI as the source of truth then the complexity around how and when Kubelet manages various obscure sync edge-cases just disappears completely. But per https://github.com/aws/amazon-vpc-cni-k8s/issues/712#issuecomment-552611994, saving that for a follow-up PR :smile: